src/gpu/kmeans/KmMatrix/KmMatrix.cpp

```diff
@@ -0,0 +1,311 @@
+/*!
```

Comment:
I can see you have implemented KmMatrix using the pimpl design pattern. We did this in xgboost for HostDeviceVector only because it had to compile for CPU only as well as with GPU. This seems unnecessarily complicated for this project. I will check with the others if we need to maintain this requirement of compiling for CPU only.

After looking through your code your matrix class seems very similar in function to https://github.com/h2oai/h2o4gpu/blob/master/src/gpu/data/matrix.cuh

Please explain why you can't use this.

Reply:
For stater, the matrix in data doesn't deal with CPU and expanding it to support CPU is just rewrite. But now that dropping CPU support completely is possible, this seem avail.

Device context is an explicit parameter for every function, but not all functions use it. It's ok if we just want a matrix that works for current implementation. But if we do want to make a generic data structure that can encapsulate the complexity, the matrix in data might also need a big change.

Surely, with all the time I spent on writing KmMatrix, the above can be done. The real difficulty is, there is no unit test in CPP, I need to rely on python tests which are more of integration tests. I tried to run the complete python tests from master branch on my machine and only a fraction of them passed. After that I dropped the thought of relying on existing tests.

I hope we can discuss this more lengthily.

Reply:
I don't actually mind which direction you come at it from but in the end these two matrix representations must be unified. I think we still want to support matrices with CPU memory, but it is no longer a requirement that the code must compile without cuda.

> Device context is an explicit parameter for every function, but not all functions use it.

You could possibly remove this and rely on your Singleton class to get things like the cublas handle inside these functions.

> The real difficulty is, there is no unit test in CPP, I need to rely on python tests which are more of integration tests. I tried to run the complete python tests from master branch on my machine and only a fraction of them passed. After that I dropped the thought of relying on existing tests.

This is worrying. Perhaps it's a difference in your environment. We will see when we sort out continuous integration.